### ISSUE TITLE
Memory leak after calling RemoteFunction

### ISSUE THREAD

**Issue by LightTab2** [DarthTon/Blackbone#283]
Hi!
Following code is almost the same as the one in the samples. The problem is it leads to a memory leak.
```cpp
using namespace blackbone;
int main()
{
	POINT point;
	Process proc;
	proc.Attach(Process::EnumByName(L"notepad.exe").front());
	proc.remote().CreateRPCEnvironment(Worker_None, true); //changing parameters there doesn't help/leads to a crash
	for (;;)
	{
		auto func = MakeRemoteFunction<decltype(&GetCursorPos)>(proc, L"user32.dll", "GetCursorPos");
		func(&point);

		//these doesn't help
		//proc.remote().reset();
		//proc.remote().memory().reset();
		//proc.memory().reset();
		Sleep(101);
	}
}
```

I've also tried to make an assembly call (i think it's named such):
```cpp
//...
auto asmPtr = AsmFactory::GetAssembler();
auto& a = *asmPtr;

a.GenPrologue();
//changing AsmVariant::eType leads to a crash, changing eCalligConvention to cc_fastcall or cc_thiscall doesn't help
a.GenCall(proc[i].modules().GetExport(L"user32.dll", "GetCursorPos")->procAddress, {&point});
//a->popa(); //crashes
a.GenEpilogue();

for(;;)
{
//...
uint64_t result = 0;
//ExecInAnyThread doesn't help, changing eThreadChangeSwitch to ForceSwitch crashes
proc[i].remote().ExecInNewThread(a->make(), a->getCodeSize(), result);
}
```

**Comment by DarthTon** [comment]
Apparently `NtTerminateThread` used in `ExecInNewThread` deadlocks for some reason. I'll look into it. `ExecDirect` works just like `CreateRemoteThread`, so it doen't support more than one arg.

**Comment by DarthTon** [comment]
In my case they don't. `_threads` should only affect source process and not the target, but they indeed do hold handles for terminated threads. I'll probably redesign it

**Comment by LightTab2** [comment]
Well I made a short-term investigation, so they probably wasn't cleaned/dealt with. I think I've got a main lead to solve this.

NTSTATUS RemoteExec::CopyCode(PVOID pCode, size_t size) in RemoteExec.cpp
marks _usercode as invalid everytime(or almost) it is executed. It allocates new memory and replaces _userCode variable, probably not cleaning before, because something invalidated it and it cannot be cleaned.

**Comment by LightTab2** [comment]
Hey did you really mean
`_assembler.push( 0 );`
in 
`void AsmHelper32::ExitThreadWithStatus( uint64_t pExitThread, uint64_t resultPtr )`
?
I think it was mean to be 
`_assembler.mov(asmjit::host::eax, 0 );`
Memory leak is solved by this small fix.

**Comment by LightTab2** [comment]
Hey did you really mean
`_assembler.push( 0 );`
in 
`void AsmHelper32::ExitThreadWithStatus( uint64_t pExitThread, uint64_t resultPtr )`
?
I think it was mean to be 
`_assembler.mov(asmjit::host::eax, 0 );`
Memory leak is solved by this small fix. However, ExecInNewThread still fails.

**Comment by LightTab2** [comment]
Ok I've got this, issue can be closed. 
`a->pop(asmjit::host::eax);` needs to be called after 
`void AsmHelper32::GenCall( const AsmFunctionPtr& pFN, const std::vector<AsmVariant>& args, eCalligConvention cc /*= CC_stdcall*/ )` to prevent memory leak in "assembly call". I am pretty sure it can be added at the bottom of this function. Will you consider this?

**Comment by DarthTon** [comment]
I don't see how this fixes the leak. IMO it just imbalances the stack leading to eventual crash. Unit tests clearly show this. 
As for `_assembler.push( 0 );` it's intended. It's a first argument for `NtTerminateThread` call

**Comment by DarthTon** [comment]
Heap debugging doesn't show any obvious leak. One candidate is `asmjit::VMemMgr::MemNode` but I dnon't know when that memory is released. `Unresolved allocations` is mostly ntdll internal `LdrAllocateTls`/`RtlProcessFlsData`, nothing that I can fix.
![image](https://user-images.githubusercontent.com/2190385/43707995-03dc3b9e-9972-11e8-9910-0b7094a06ea2.png)

**Comment by DarthTon** [comment]
You can run tests locally or recreate a PR. CI was just running tests before you closed it and tests were cancelled.

**Comment by LightTab2** [comment]
It seems issue isn't related with GenCall directly, following code prevents memory leak as well.
```cpp
    a.GenPrologue();
	a.GenCall(0x450AC0, { 3,11,11,0,0 });
	a.GenCall(0x450AC0, { 3,-10,-10,0,0 });
	a->pop(asmjit::host::eax); //"Fixes" memory leak for both GenCalls
	a.GenEpilogue();
	proc[i].remote().ExecInNewThread(a->make(), a->getCodeSize(), result);
```

I'll leave it looped for half an hour and tell you if some fatal corruption occured.

**Comment by LightTab2** [comment]
No crash, though...
![th](https://user-images.githubusercontent.com/22819912/43710899-52d7c228-9971-11e8-9cb7-39f94b5cbe7a.png)
> In my case they don't.
Was it accordingly to threads being terminated or _threads?

Cuz this one is an issue, slowing down entire process.

**Comment by DarthTon** [comment]
> In my case they don't.
I was wrong. Messed up some code during tests. `_threads` is easily fixable tho

**Comment by LightTab2** [comment]
Yeah, however fixing _threads isn't the only memory leak in the calling process (executing ExecInNewThread). I am into it currently.

**Comment by LightTab2** [comment]
_threads aren't the only memory leak. Seems like a->make() passed as a parameter to ExecInNewThread needs to be cleaned too.

**Comment by LightTab2** [comment]
It isn't the only one memory leak. Seems like memory allocated in _runtime in Assembler isn't freed anywhere (maybe it's done in includes or assembly, but I can't find the call to release() function).

**Comment by LightTab2** [comment]
![obraz](https://user-images.githubusercontent.com/22819912/43779471-c55fcc86-9a58-11e8-8630-a9350c33d4bd.png)
